src/util-bpf.c

```diff
@@ -75,3 +75,276 @@ int SCBPFCompile(int snaplen_arg, int linktype_arg, struct bpf_program *program,
+ * BPF.
+ * 
+ * \param argc Number of command line arguments.
+ * \param argv Actual command line arguments.
+ *
+ * \retval TM_ECODE_OK will be returned if the BPF is set, otherwise
+ *   TM_ECODE_FAILED will be returned.
+ */
+int SetBpfString(int argc, char *argv[])
+{
+    char *bpf_filter = NULL;
+    uint32_t bpf_len = 0;
+    int tmpindex = 0;
+
+    /* attempt to parse remaining args as bpf */
+    tmpindex = argc;
+    while(argv[tmpindex] != NULL) {
+        bpf_len+=strlen(argv[tmpindex]) + 1;
+        tmpindex++;
+    }
+
+    if (bpf_len == 0)
+        return TM_ECODE_OK;
+
+    if (EngineModeIsIPS()) {
+        SCLogError(SC_ERR_NOT_SUPPORTED,
+                   "BPF not available in IPS mode."
+                   " Use firewall filtering if possible.");
+        return TM_ECODE_FAILED;
+    }
+
+    bpf_filter = SCMalloc(bpf_len);
+    if (unlikely(bpf_filter == NULL))
+        return TM_ECODE_OK;
+    memset(bpf_filter, 0x00, bpf_len);
+
+    tmpindex = optind;
+    while(argv[tmpindex] != NULL) {
+        strlcat(bpf_filter, argv[tmpindex],bpf_len);
+        if(argv[tmpindex + 1] != NULL) {
+            strlcat(bpf_filter," ", bpf_len);
+        }
+        tmpindex++;
+    }
+
+    if(strlen(bpf_filter) > 0) {
+        if (ConfSetFinal("bpf-filter", bpf_filter) != 1) {
+            SCLogError(SC_ERR_FATAL, "Failed to set BPF.");
+            SCFree(bpf_filter);
+            return TM_ECODE_FAILED;
+        }
+    }
+    SCFree(bpf_filter);
+
+    return TM_ECODE_OK;
+}
+
+/**
+ * \brief Set the BPF based on the specified file.
+ *
+ * Make use of the ParseBpfFromFile function to parse the BPF and sets
+ * the "bpf-filter" configuration option with the result.
+ * 
+ * \param filename Name of the file containing the BPF.
+ *
+ * \retval TM_ECODE_OK will be returned if the BPF is parsed, otherwise
+ *   TM_ECODE_FAILED will be returned.
+ */
+int SetBpfStringFromFile(const char *filename)
+{
+    const char *bpf_filter = NULL;
+
+    int ret = ParseBpfFromFile(filename, &bpf_filter);
+
+    if (ret == TM_ECODE_OK) {
+        if(ConfSetFinal("bpf-filter", bpf_filter) != 1) {
+            SCLogError(SC_ERR_FOPEN, "ERROR: Failed to set the BPF!");
+            ret = TM_ECODE_FAILED;
+        }
+    }
+    
+    if (bpf_filter != NULL) {
+        SCFree((char *)bpf_filter);
+    }
+
+    return ret;
+}
+
+/**
+ * \brief Parse the BPF from the Suricata configuration file.
+ *
+ * This function parses the "bpf-filter" configuration option and
+ * supports files as well as BPF expressions. Two notes, first the command
+ * line value has precedence, second the pointer retuned in vptr needs to be
+ * freed by the caller in case of success.
+ * 
+ * \param vptr Pointer that will be set to the configuration value parameter.
+ *
+ * \retval 1 will be returned if the BPF is parsed, otherwise
+ *   0 will be returned.
+ */
+int ParseBpfConfig(ConfNode *if_root, char **vptr)
```

Comment:
is there a reason to have an `int` retval in addition to returning through `vptr`? Seems returning `vptr` or `NULL` in case of failure would be simpler.